### DISCUSSION TITLE
Problems on customizing high-level HTTP Client for requesting upstream (sharing experience as well as Q&A)

### DISCUSSION THREAD

**Discussion by cxw620** [hyperium/tonic#1631] [Q&A]
So far `Tonic` do not support proxy, so I have to write a custom HTTP client and use with `EchoClient::with_origin(client, uri)` (take `EchoClient` as an example which's generated with tonic-build).

At first, I simply use `hyper::Client` with my customized `Connector` with proxy support. However, when compression is enabled, `Tonic` will not work with error: `protocol error: received message with compressed-flag but no grpc-encoding was specified`. I've checked docs and found nothing related with such error and I have to read source codes.

- Func call tracking:
  https://github.com/hyperium/tonic/blob/e31f5ccfc339e9f3d82a020e28751f188c81c564/tonic/src/client/grpc.rs#L280-L367

  `stream` ->-> `create_response` -> expect_additional_trailers = false -> `Streaming::new_empty(decoder, body)` -> `Streaming::new(decoder, body, Direction::EmptyResponse, None, None)`

  Here encoding is `None` though actually `Gzip`, then error occurs.

In func `stream`, inner T `call` the request and gets response from upstream. By comparing the response headers between offcial example and my own one, I noticed that `grpc-status`, `grpc-status-message` and `grpc-status-details-bin` were removed when in offcial example. I guessed that tonic `Channel` will handle gRPC Status before decoding response and remove `grpc-status`, `grpc-status-message` and `grpc-status-details-bin` from response headers, while hyper::Client will not do so.

Finally, I write a wrapper of hyper::Client and impl `tower::Service` for it, then check tonic::Status and remove `grpc-status`, `grpc-status-message` and `grpc-status-details-bin` from response headers. It works!

I'm not sure if I'm right and I would be really happy if anyone could share more details in the comment area.

P.S. Any convenient ways on tracing function calls in Rust?

**Comment by clouedoc** [comment]
I'm interested in using proxies with Tonic but I'm a Rust newbie and it feels like I'll end up working on it for 3 days before ending up finding weird bugs (like compression not working – thanks for mentioning it, since it's rather critical to my use case...)

Perhaps the simplest way to go about this would be to fork tonic and add the functionality there?
I'm going to pursue this approach in my fork and report here.

**Reply by clouedoc** [reply]
EDIT: I ended up patching tonic with success to support SOCKS5 proxies: https://github.com/hyperium/tonic/discussions/1206#discussioncomment-12080630

There were no issues connecting to the target service despite compression enabled, so I would guess the gRPC metadata headers were preserved. Same with TLS, I didn't need to do anything in particular despite the target service requiring mutual TLS (mTLS). In fact, I had to remove the SslConnector that I copied from your code from 2022, because it messed up with the TLS configuration I already had.

I suspect that patching Tonic is the simplest approach going forward for bugfixes or feature implementation.

A lot of thanks for your previous messages from 2022 on the other discussion; without it, it would have took me a lot of time to figure it out... It ended up taking me 2 hours instead of 3 days. Cheers :-)
